Grails-4.0.4: init fails, it can not find services

I can not start a new test application with version 4.0.4 because init doesn't find the Service.
The same with an application, which works on 4.0.3 (developer mode)

### Task List

- [X] Steps to reproduce provided
- [X] Stacktrace (if present) provided
- [ ] Example that reproduces the problem uploaded to Github
- [X] Full description of the issue provided (see below)

### Steps to Reproduce

```
grails create-app test
cd test
grails create-domain test.Book
## write in grails-app/domain/test/Book.groovy a property eg: String name
grails generate-all test.Book
grails run-app
```

### Expected Behaviour

Application should start

### Actual Behaviour

init fails because it can not find the BookService.
All the necessary files are in place

### Stacktrace
```

...
: Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'bookService': Invocation of init method failed; nested exception is org.grails.datastore.mapping.services.ServiceNotFoundException: No service found for type interface test.BookService
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1771)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:593)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:226)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:841)
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:878)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:550)
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:141)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:744)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:391)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:312)
        at grails.boot.GrailsApp.run(GrailsApp.groovy:99)
        at grails.boot.GrailsApp.run(GrailsApp.groovy:485)
        at grails.boot.GrailsApp.run(GrailsApp.groovy:472)
        at test.Application.main(Application.groovy:11)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49)
Caused by: org.grails.datastore.mapping.services.ServiceNotFoundException: No service found for type interface test.BookService
        at org.grails.datastore.mapping.services.DefaultServiceRegistry.getService(DefaultServiceRegistry.groovy:81)
        at org.grails.datastore.mapping.core.AbstractDatastore.getService(AbstractDatastore.java:102)
        at sun.reflect.GeneratedMethodAccessor64.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.util.MethodInvoker.invoke(MethodInvoker.java:280)
        at org.springframework.beans.factory.config.MethodInvokingBean.invokeWithTargetException(MethodInvokingBean.java:123)
        at org.grails.datastore.mapping.config.DatastoreServiceMethodInvokingFactoryBean.invokeWithTargetException(DatastoreServiceMethodInvokingFactoryBean.groovy:30)
        at org.springframework.beans.factory.config.MethodInvokingFactoryBean.afterPropertiesSet(MethodInvokingFactoryBean.java:108)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1830)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1767)
        ... 22 common frames omitted
```

### Environment Information

- **Operating System**:

```
$ lsb_release  -a
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.6 LTS
Release:        14.04
Codename:       trusty
$ uname -a
Linux hg1-9004-2 3.13.0-170-generic grails/grails-core#220-Ubuntu SMP Thu May 9 12:40:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

- **Grails Version:**
```
$ grails -version
| Grails Version: 4.0.4
| JVM Version: 1.8.0_201


```

- **JDK Version:**

```
$ java -version
java version "1.8.0_201"
Java(TM) SE Runtime Environment (build 1.8.0_201-b09)
Java HotSpot(TM) 64-Bit Server VM (build 25.201-b09, mixed mode)
```

### Example Application

Standard New Application with a Domain Class test.Book and all automatically generated Controller, Service, Views etc

```
package test
class Book {
	String name
	static constraints = {
	}
}
```

### Thank you for all, love Grails :-)

--- a/grails-datastore-core/src/main/groovy/org/grails/datastore/mapping/services/DefaultServiceRegistry.groovy
+++ b/grails-datastore-core/src/main/groovy/org/grails/datastore/mapping/services/DefaultServiceRegistry.groovy
@@ -27,14 +27,14 @@ class DefaultServiceRegistry implements ServiceRegistry, Initializable {
      */
     final Datastore datastore
 
-    protected final Map<Class,Service> servicesByInterface
+    protected final Map<String,Service> servicesByInterface
     protected final Collection<Service> services = []
     private boolean initialized
 
     DefaultServiceRegistry(Datastore datastore, boolean exceptionOnLoadError = true) {
         this.datastore = datastore
         Iterable<Service> services = loadServices()
-        Map<Class,Service> serviceMap = [:]
+        Map<String, Service> serviceMap = [:]
         Iterator<Service> serviceIterator = services.iterator()
         while(serviceIterator.hasNext()) {
             try {
@@ -42,18 +42,18 @@ class DefaultServiceRegistry implements ServiceRegistry, Initializable {
                 this.services.add(service)
                 Class[] allInterfaces = ClassUtils.getAllInterfaces(service)
                 Class theClass = service.getClass()
-                serviceMap.put(theClass, service)
+                serviceMap.put(theClass.name, service)
                 if( theClass.simpleName.startsWith('$') ) {
                     // handle automatically implemented abstract service implementations
                     Class superClass = theClass.getSuperclass()
                     if(superClass != null && superClass != Object.class && Modifier.isAbstract(superClass.modifiers)) {
-                        serviceMap.put(superClass, service)
+                        serviceMap.put(superClass.name, service)
                     }
 
                 }
                 for(Class i in allInterfaces) {
                     if(isValidInterface(i)) {
-                        serviceMap.put(i, service)
+                        serviceMap.put(i.name, service)
                     }
                 }
             } catch (Throwable e) {
@@ -76,7 +76,7 @@ class DefaultServiceRegistry implements ServiceRegistry, Initializable {
 
     @Override
     def <T> T getService(Class<T> interfaceType) throws ServiceNotFoundException {
-        Service s = servicesByInterface.get(interfaceType)
+        final Service s = servicesByInterface.get(interfaceType.name)
         if(s == null) {
             throw new ServiceNotFoundException("No service found for type $interfaceType")
         }

